[#47] 주문 생성 기능 구현

chicori3 opened the pull request:
## 목표

유저는 장바구니에 추가한 강의 중 하나 이상을 선택하여 주문을 생성할 수 있다

## 할 일

- [x] 장바구니에 없는 강의인 경우 예외가 발생한다
- [x] 유저는 사용가능한 쿠폰을 적용할 수 있다
  - [x] 쿠폰이 유효하지 않을 경우 예외가 발생한다
- [x] 생성된 주문 정보에는 주문 상태, 주문 일자, 주문한 강의 정보, 총 결제 금액, 할인 금액이 있어야 한다

## 고려사항

- 장바구니와 장바구니 아이템을 한 번에 조회하도록 fetch join을 사용했습니다.
- 애플리케이션 계층 통합테스트를 위해 IntegrationTest를 작성했습니다.
- 통합테스트를 위한 샘플 데이터를 반복 작성하지 않도록 Test Fixture를 작성했습니다.

resolve: #47

chicori3 committed:
refactor(#6) - Coupon 도메인 수정, 에러 코드 추가, 메서드 참조 식으로 변경

chicori3 committed:
feat(#6) - Order 도메인 작성

chicori3 committed:
feat(#6) - BaseEntity 작성, JpaAuditing 설정

chicori3 committed:
feat(#6) - 주문 생성 리포지토리 코드 작성

chicori3 committed:
feat(#6) - 주문 생성 서비스 계층 작성

chicori3 committed:
refactor(#6) - 쿠폰 리포지토리 Stub 네이밍 변경, DiscountType 할인값을 BigDecimal 타입으로 변경

chicori3 committed:
refactor(#6) - 사용하지 않는 CouponProcessor 삭제

chicori3 committed:
feat(#6) - 장바구니에 해당 강의가 있는지 검증하는 로직 추가

chicori3 committed:
feat(#6) - 회원 Role을 Teacher로 승급하는 로직 추가

chicori3 committed:
feat(#6) - 테스트를 위한 Fixture 작성

chicori3 committed:
refactor(#6) - 도메인 로직 수정, OrderItem 엔티티 작성, 에러 코드 추가

chicori3 committed:
refactor(#6) - 주문 생성 로직 수정

chicori3 committed:
feat(#6) - 통합 테스트 기본 클래스와 DB 초기화 객체 작성

chicori3 committed:
test(#6) - 주문 생성 테스트 작성

chicori3 committed:
chore(#6) - application.yaml 수정

chicori3 committed:
feat(#6) - 주문 생성 컨트롤러 작성

chicori3 committed:
test(#6) - 주문 생성 컨트롤러 테스트 작성

chicori3 committed:
refactor(#6) - 주문한 강의 정보 추가

chicori3 committed:
refactor(#6) - Cart 도메인 수정

- CartItem 엔티티 승격
- Cart-CartItems 일대다 연관관계 설정
- Fixture 수정

chicori3 committed:
refactor(#6) - Cart 리포지토리 수정

- CartItem 까지 조회하는 join fetch 쿼리 추가

chicori3 committed:
refactor(#6) - Cart 서비스 계층 수정

- Cart 도메인 수정으로 인한 변경

f-lab-michael commented on src/main/java/com/flab/infrun/coupon/domain/Coupon.java:
```diff
@@ -104,14 +107,31 @@ private void verifyIsRegistrable(final LocalDateTime currentTime) {
+        return this.discountInfo.discount(price);
+    }
+
+    public void isUsable(final LocalDateTime currentTime, final Member customer) {
```
isUsable이면 boolean을 리턴할거 같은 함수 이름이라 check같은 문구로 시작하는게 더 좋을것 같습니다.

f-lab-michael commented on src/main/java/com/flab/infrun/coupon/domain/Coupon.java:
```diff
@@ -104,14 +107,31 @@ private void verifyIsRegistrable(final LocalDateTime currentTime) {
         this.status = CouponStatus.USED;
     }
 
+    public void rollback() {
```
rollback은 원래 상태로 돌린다는 의미일텐데 이 경우는 원래 상태 라기 보다 무조건 REGISTERED로 변경하는거라 다른이름이 좋지 않을까요?

chicori3 replied:
주문 프로세스 중에 실패하는 경우 쿠폰의 상태를 되돌리기 위해 rollback이라고 명시했는데 
쿠폰 상태 이외에 필드들도 rollback의 대상이 될 수도 있으니 말씀해주신대로 적절하지 않은 것 같습니다

f-lab-michael commented on src/main/java/com/flab/infrun/coupon/domain/DiscountInfo.java:
```diff
@@ -9,22 +10,28 @@
 
     private DiscountType discountType;
-    private int discountValue;
+    private BigDecimal discountValue;
```
BigDecimal 로 바꾼 이유가 있을까요?

chicori3 replied:
가격 관련한 필드들을 BigDecimal 타입으로 바꾸면서 함께 바꿨는데
다시 생각해보니 discountValue는 할인율 또는 할인가를 의미하고 두 값 모두 정수형으로 표시할 수 있으니 적절하지 않은 것 같습니다. 😅

f-lab-michael commented on src/main/java/com/flab/infrun/order/application/CreateOrderProcessor.java:
```diff
@@ -0,0 +1,98 @@
+        return CreatedOrderResult.from(order);
+    }
+
+    private Coupon verifyCoupon(
```
coupon을 리턴하니 getValidatedCoupon 이라던지 vatifyAndGetCoupon 이라던지 함수명을 좀더 명시적으로 사용하는게 좋을것 같아요

chicori3 committed:
refactor(#6) - 피드백 반영

- 부적절한 메서드 네이밍 수정
- DiscountValue의 타입을 BigDeicmal -> int로 변경
- 주문 생성 시 쿠폰을 확인하고 반환할 때 Optional 타입으로 반환

chicori3 committed:
refactor(#6) - DatabaseCleaner가 작동하도록 @Table 어노테이션 추가

chicori3 committed:
test(#6) - 주문 생성 테스트 케이스 추가 작성
